oauth-proxy/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,159 @@
+side-cars.  Because it is re-used by directly embedding the `oauth-proxy` into a `PodSpec`, any changes to the way it is
+invoked will cause adoption difficulties.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions
+
+1. Can we expose proxy.config.openshift.io to system:authenticated by default.  Some cluster-admins may place basic-auth
+ credentials here because we provide no alternative.  This may make the configuration values suddenly sensitive and there
+ will be desire to hide that from workloads and other users.
+
+## Summary
+
+In order to function, the `oauth-proxy`
+ 1. Gets the oauth-server location using the kube-apiserver's internal IP: `https://172.30.0.1/.well-known/oauth-authorization-server`
+ 2. Uses the result to find the token endpoint.
+ 3. Hits the *external* token endpoint `https://oauth-openshift.apps.<cluster-dns-prefix>.openshift.com/oauth/token`.  
+
+This flow requires hitting a potentially external router, which requires honoring proxy settings.  `/.well-known/oauth-authorization-server`
+implements [RFC 8414](https://tools.ietf.org/html/rfc8414) which does not allow specifying multiple possible URLs.
+
+## Motivation
+
+See summary in first section.
+
+### Goals
+
+1. Zero-config proxy support that dynamically updates as proxy settings change.
+2. Support for all proxy modes, including MITM.
+
+### Non-Goals
+
+1. Making generic code to rewrite proxy transport configuration.
+
+## Proposal
+
+The `oauth-proxy` can be written to use targeted list/watch against the proxy.config.openshift.io and 
+`oc get -n openshift-config-managed configmaps/trusted-ca-bundle` to build a proxy transport.  Using standard informers with single item
+watches allows RBAC to only expose GET for individual items.  It also means that an efficient observed update and `sync.Atomic` mechanism can be built
+to ensure performant behavior.
+
+This is an intricate thing to write, so we avoided it for every other proxy component, but this one is a component that is
+re-used by numerous external teams by directly embedding it into a pod spec.  This changes the tradeoff, so it becomes worthwhile
+to write this code.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. If HTTP env vars are set, the dynamic proxy-support code must be disabled.  The user is in charge of what the `oauth-proxy`
+ does in these cases.
+2. The list/watch driving the informers must be focused to a single item, so the RBAC special case for single item watch
+ takes effect.
+3. To be performant, the proxy and TLS configuration should be cached in something like a `sync.Atomic`.
```

Comment:
proxy setting nearly never hange. Why must this be performant?

Reply:
> proxy setting nearly never hange. Why must this be performant?

The loading needs to be performant. Using normal locks isn't as good for mostly-read load I don't think